### DISCUSSION TITLE
The case for menu bar access

### DISCUSSION THREAD

**Discussion by Ptujec** [Ptujec/LaunchBar#19] [Ideas]
When I first came across actions/extension that let you access menubar items via a launcher I didn’t really get it. Where is the added benefit compared to the built in search?
With some of the solutions I really didn’t see any. This however changed dramatically a while ago. 
I can’t really remember how and why, but I tried it again, when I updated my LaunchBar setup with the purchase of my MBA M1. Or maybe it was related to Raycast advertising it as a feature. 

## Here is why I’m using it all the time now: 

1. The version I use is really **fast**, almost instant. 
2. Instead of using it like a standard search I made it so it lists all items right away. I then type to narrow results.

Both of those factors are important. The first one should be obvious. The main benefit of the second is that it utilizes the superpower of LB. You’ll find stuff more easily without knowing the exact name. It can be **triggered simply by space** (which is convenient because my thumb is still hovering over it after bringing up LaunchBar). Plus **LB learns and remembers your preferences**. So every time I press `P` in Pages it’ll open the PDF export dialogue. `C` in Pixelmator Pro opens the canvas size dialogue, …

I absolutely love it. No need to set up and remember custom shortcuts.

## Why is this a discussion?

I feel LB users would really benefit from this kind of functionality, but currently there is no action for it that works. Unfortunately I can not share mine, because it is based on an executable, which is originally part of an  [Alfred Workflow](https://github.com/BenziAhamed/Menu-Bar-Search). You would be running into the same [issue](https://github.com/jordicl/launchbar-menubar-search/issues/5) I did with [Jordi's action](https://github.com/jordicl/launchbar-menubar-search). 
I don’t feel like paying 100$ each year to sign my actions to share them for free.

[Roey Biran](https://github.com/roeybiran) has a dedicated app for this called [Finbar](https://www.roeybiran.com/apps/finbar). I would be more then happy to support it if he could offer an integration that would work as described above. Or this functionality could be directly integrated by Obdev. 

What do you think? Would you use something like that? Questions?

**Comment by hodgesd** [comment]
Sounds amazing.  

I've been a LaunchBar user for a decade.  I would happily pay Obdev for an upgrade or support Roey get this feature.

**Comment by andrewsardone** [comment]
As an avid  <kbd>⇧ ⌘ /</kbd> user to search and engage menu bar actions, I would absolutely prefer to use LaunchBar. For all the reasons you described above, @Ptujec, LaunchBar is a better tool.

### What's the next step?

1. @Ptujec, is it possible for your to document how your current menu-bar actions workflow works? I understand the signing hiccup, but perhaps for the time being individual users could take on that burden.
2. Is there a feature request to LaunchBar I can +1? It would be fantastic if I could search the menu bar natively in LaunchBar [the way I can other categories](https://user-images.githubusercontent.com/14554/220170615-8da91f2e-9c53-41b9-885e-3f7497565dcb.png).

**Reply by Ptujec** [reply]
I [forked](https://github.com/Ptujec/Menu-Bar-Search) the source code of the Alfred Workflow executable and made a minor tweak to get title and shortcuts separately (see my screenshot). Here is what I do with the json data that the executable produces: https://gist.github.com/Ptujec/80aa113794d8fba1d5ac5a5caffd3ef1

**Reply by Ptujec** [reply]
@mlinzner is also on board. https://mastodon.nl/@ptujec@hessen.social/109897605461202518
I am sure Obdev will also take note. They favorited almost all my LaunchBar related stuff on Mastodon 😁. Not sure if they will comment though. They don't seem to make feature promises.

**Comment by gramata** [comment]
Thanks for pointing out Finbar hadn't come across that yet. 

I currently use Paletro.app which feels very much like LaunchBar for the Menu Bar to me. My shortcut being `⌘ ⌃ /`

I have tried out the Menu Bar Search action by Edward Cheng that is here: https://github.com/ewlcheng/launchbar-actions
It works well. @Ptujec Perhaps you can have a look at what Edward is doing if you are having issues with your action. 

It's great to see the renewed interest in the LaunchBar like app space.
This feature is also a plug-in for Raycast which I've tried as well. 
A handy feature that would be great to be adopted by LaunchBar is to be able to add shortcuts to trigger actions. 
Eg: I have `⌘ ⌥ /`  for jumping to Raycast Menu Bar Search plugin directly.

Thank you for all the great work you have put into your actions. They add to LaunchBar being indispensable to me. 
While I've added Raycast to the toolset it is very different conceptually to the search and manipulation of text and files that LaunchBar allows and I could not live without.

**Reply by Ptujec** [reply]
Thank you for reminding me about Edward! I think his action was actually what sparked my renewed interest. I knew there was another action, but I could not find it. The thing is, I even [contributed](https://github.com/ewlcheng/launchbar-actions/commit/6097711932f1f2c0feea4a842d160012dfd87def) to his action to make it work as described above. I should have checked my Twitter DMs. 🤦🏻‍♂️ 

Ultimately it was just a little bit to slow for me. It is not Eds fault. [The AppleScript that is pulling the menu bar items is the bottleneck](https://github.com/ewlcheng/launchbar-actions/blob/5f9a1499af4569fbf09253afa2e3cc1c50b02ba5/Search%20Menu%20Bar.lbaction/Contents/Scripts/default.js#L7).
I looked for ways to improve speed and came across the Alfred workflow. But Ed (like me) didn't feel comfortable sharing an action with an executable that will give you a malware warning (even though there is ways you can turn off the warning). 

I enjoy sharing the tools, that I love and I can't really share this one. This is the only problem. Good problem to have, huh 😁. 

So I think the best solution is probably Objective Development adding this as a feature. No security issues, no need to install anything, speed, and most importantly all the benefits from the LaunchBar way of doing things. I tested a few other solutions but none can do what LB does. Not everyone - not even every LB user - cares about what makes this special to me (as described above). That's fine. But I bet I am not the only one.

**Reply by Ptujec** [reply]
BTW: In the testing of the Raycast integration I found that it would miss some items, e.g. it can not find the PDF export item in Pages.

**Comment by roeybiran** [comment]
Hi! My name’s Roey and i’m the developer behind [Finbar](https://www.roeybiran.com/apps/finbar). Integrations with 3rd party apps such as LaunchBar are already in the works and hope to add thme in next version or two.

**Reply by Ptujec** [reply]
Looking forward to giving it a try.

**Reply by Ptujec** [reply]
It might already be clear to you but let me point out again what I like about my action and what I would look for in an integration. Finbar remembers used items. This is great! It would actually be a great addition to my action (**Update**: I just added it and also updated my [gist](https://gist.github.com/Ptujec/80aa113794d8fba1d5ac5a5caffd3ef1)). 
However it does not have the same filtering capabilities LB has. In LB I type "PDF" only the first time. After selecting the item once, I don't have to type "PDF" the next time. It already has the menu item selected ready to launch when I start typing "P".

**Reply by Ptujec** [reply]
Also, props on the speed!

**Reply by mlinzner** [reply]
Nice work @roeybiran, just bought a license and looking forward to the integrations. I'm happy to take care of the LaunchBar action side once you have something to look at and I'm pretty sure @Ptujec would be happy to assist as well ;)

**Reply by roeybiran** [reply]
Thank you so much! Hopefully I’ll release it this weekend as I’m busy with my front-end developer day job. Luckily making Mac apps is just way too fun 😁

**Reply by mlinzner** [reply]
Tell me about it, we are all in the same boat 😂

**Comment by roeybiran** [comment]
A week behind schedule, but as of Finbar v1.3 (released today), a command line tool named `finbar-cli` is now included with the app. The tool offers programmatic access to some of Finbar’s basic functionalities. Of course, this is just the first version of tool and I’m open to any feedback and suggestions. In fact, you can already expect some nice performance improvements very soon.

Read more in [Finbar’s user guide](http://roeybiran.com/apps/finbar/help).

Enjoy!

**Reply by andrewsardone** [reply]
That's awesome, @roeybiran! [An instant purchase](https://user-images.githubusercontent.com/14554/222929318-85226546-eb53-4e1f-bbc8-22e4a8d68c6b.png) 🎉 

At some point I'll dig into shelling out to `finbar-cli` for a LaunchBar, but for the time being I'm digging just using the standard <kbd>⇧ ⌘ /</kbd> shortcut to invoke Finbar.

**Reply by Ptujec** [reply]
This is how far I got. But the clicking does not work. Maybe @mlinzner has a hint for me? ;) 


```
/* 
Menu Bar Items (powered by Finbar) Action for LaunchBar
by Christian Bender (@ptujec)
2023-03-04

Copyright see: https://github.com/Ptujec/LaunchBar/blob/master/LICENSE

TODO: 
- click
- recent items
*/

const finbarCLI = '/Applications/Finbar.app/Contents/MacOS/finbar-cli';

function run(argument) {
  //
  var finbarList = LaunchBar.execute(finbarCLI, 'list');

  var data = JSON.parse(finbarList);

  var result = [];
  data.forEach(function (item) {
    var sub = item.path.join(' > ');
    var uid = item.bundleID + '>' + item.path.join('>');

    if (item.indices[0] != 0 && !item.path.includes('Open Recent')) {
      result.push({
        title: item.title,
        subtitle: sub,
        icon: item.bundleID,
        action: 'click',
        actionArgument: item,
        actionRunsInBackground: true,
      });
    }
  });

  return result;
}

function click(item) {
  LaunchBar.hide();
  // For example, to click the “About This Mac” menu item of the “Apple” menu bar item, specify the following arguments to the `select` command: `0 1 --title="About This Mac"`.

  var title = '--title="' + item.title + '"';
  var indices = item.indices.join(' ');
  // LaunchBar.alert(finbarCLI + ' select ' + indices + ' ' + title);

  LaunchBar.execute(finbarCLI, 'select', indices, title);
}
```

**Reply by mlinzner** [reply]
Awesome work @roeybiran! Can't wait to give it a try. 

@Ptujec I'm afk this weekend and the LaunchBar docs are not loading for me. But from what I remember, JavaScript actions should reference the open action as function object and not string. So I would give it another try after removing the quotes around click.

**Comment by roeybiran** [comment]
Hi folks, thank you so much for purchases and support — it’s truly heartwarming.

@Ptujec, I scaffolded an LB action for Finbar which seems to work really well, I’d appreciate it if you’d give it a shot.

```js
// LaunchBar Action Script

const finbarCLI = "/Applications/Finbar.app/Contents/MacOS/finbar-cli";

function run() {
  return JSON.parse(LaunchBar.execute(finbarCLI, "list")).map(
    ({ title, path, bundleID, indices }) => ({
      title,
      subtitle: path.join(" > "),
      icon: bundleID,
      action: "click",
      actionArgument: { indices, title },
    })
  );
}

function click({ indices, title }) {
  LaunchBar.execute(finbarCLI, "select", ...indices, "--title", title);
}
```

If  it works for you, I’ll open up a pull request.

Sadly, between work and working on Finbar I don’t have much time left for working on LB actions 🤣 So once we’ll be satisfied from this action I’ll make an “official” integration on Finbar’s home page (similar to [what Dash does](https://kapeli.com/dash/#integration)).

P. S. — telling from your code snippet I can already see I need to add a flag to `finbar-cli` to respect any rules defined in Finbar’s UI. stay tuned!

**Reply by Ptujec** [reply]
Thanks! It works great. This shows how little I know about coding. 🤦🏻‍♂️ The spread operator (just googled that) was the missing piece I needed for LaunchBar.execute to work. 

There are two things I would like to integrate: 

1) Store the used item in Action.preferences.recentItems and show it at the top of the list if existing (per app). I can do that. 

2) An option to exclude certain items. But you found out about that already. 

BTW: The performance is great already. But improvements in that area are always welcome I guess.

Great work!

**Reply by Ptujec** [reply]
Showing keyboard shortcuts of items (if existing) would be nice at some point. It's not essential for the action but nice for "educational" purposes.

**Reply by roeybiran** [reply]
It’s there! This made me realize I need to better document the response from the `list` command.

I’ll make it available under `list -h` soon, but here it is for now:

```
title: String
path: [String]
indices: [Int]
shortcut: String?
mark: String?
bundleID: String?
```

Mark is the the little character that prefixes some menu items. Usually **✓**.

**Reply by Ptujec** [reply]
My bad. Totally missed it. Thanks

**Reply by Ptujec** [reply]
Ok. Action is up here: https://github.com/Ptujec/LaunchBar/tree/master/Menu-Bar#readme

It shows marks and shortcuts and puts the last used item on top of the list. Not sure if its better to have the last used item as a flag in the CLI to keep the code cleaner. I could add that in an update along with the rules flag.

**Comment by roeybiran** [comment]
The fault is actually mine (and the missing docs).

**Comment by Ptujec** [comment]
Action powered by Finbar is up here: https://github.com/Ptujec/LaunchBar/tree/master/Menu-Bar#readme

**Comment by roeybiran** [comment]
Amazing, thank you. I’ll link to it in Finbar’s user guide soon.

**Comment by jonohunt** [comment]
Love this. Thank you all for your combined effort! 🎉

**Comment by roeybiran** [comment]
Hi folks, sorry for taking a while, but today I released version 1.5 of my app. You can now exclude menu items directly from the command line with the `--excluding-title` option. Contrary to what I planned earlier, I’ve decided to make `finbar-cli` independent of anything that’s configured in the UI, including rules. 

As always, I encourage you to report any issues or request new features through [Finbar’s issue tracker](https://github.com/roeybiran/finbar-issues/issues), which I monitor closely.

Thanks and enjoy!
